cloudinit/sources/DataSourceVMware.py

```diff
@@ -0,0 +1,816 @@
+
+    # If there is a default IPv4 address but not IPv6, then see if there is a
+    # single IPv6 address associated with the same device associated with the
+    # default IPv4 address.
+    if ipv4 and not ipv6:
+        af_inet6 = addr4_fams.get(netifaces.AF_INET6)
+        if af_inet6:
+            if len(af_inet6) > 1:
+                LOG.warning(
+                    "device %s has more than one ipv6 address: %s",
+                    dev4,
+                    af_inet6,
+                )
+            elif "addr" in af_inet6[0]:
+                ipv6 = af_inet6[0]["addr"]
+
+    # If there is a default IPv6 address but not IPv4, then see if there is a
+    # single IPv4 address associated with the same device associated with the
+    # default IPv6 address.
+    if not ipv4 and ipv6:
+        af_inet4 = addr6_fams.get(netifaces.AF_INET)
+        if af_inet4:
+            if len(af_inet4) > 1:
+                LOG.warning(
+                    "device %s has more than one ipv4 address: %s",
+                    dev6,
+                    af_inet4,
+                )
+            elif "addr" in af_inet4[0]:
+                ipv4 = af_inet4[0]["addr"]
+
+    return ipv4, ipv6
+
+
+# patched socket.getfqdn() - see https://bugs.python.org/issue5004
+
+
+def getfqdn(name=""):
+    """Get fully qualified domain name from name.
+    An empty argument is interpreted as meaning the local host.
+    """
+    name = name.strip()
+    if not name or name == "0.0.0.0":
+        name = socket.gethostname()
+    try:
+        addrs = socket.getaddrinfo(
+            name, None, 0, socket.SOCK_DGRAM, 0, socket.AI_CANONNAME
+        )
+    except socket.error:
+        pass
+    else:
+        for addr in addrs:
+            if addr[3]:
+                name = addr[3]
+                break
+    return name
+
+
+def is_valid_ip_addr(val):
+    """
+    Returns false if the address is loopback, link local or unspecified;
+    otherwise true is returned.
+    """
+
+    addr = None
+    try:
+        addr = ipaddress.ip_address(val)
+    except ipaddress.AddressValueError:
+        # If this is Python3 then alias the built-in "str" to the old
+        # "unicode" built-in. This bypasses the Flake8 warning for using
+        # an unsupported keyword.
+        if sys.version_info[0] >= 3:
+            unicode = str
+        addr = ipaddress.ip_address(unicode(val))
+    except Exception:
+        return None
+
+    if addr.is_link_local or addr.is_loopback or addr.is_unspecified:
+        return False
+    return True
+
+
+def get_host_info():
+    """
+    Returns host information such as the host name and network interfaces.
+    """
+
+    host_info = {
+        "network": {
+            "interfaces": {
+                "by-mac": collections.OrderedDict(),
+                "by-ipv4": collections.OrderedDict(),
+                "by-ipv6": collections.OrderedDict(),
+            },
+        },
+    }
+
+    hostname = getfqdn(socket.gethostname())
```

Comment:
Hmm, I don't think it would? The logic paths are different. Elsewhere you remarked:

>  I'm used to seeing cloud-init retrieving the network information from the datasource to apply to the machine and not the other way around, so this is a bit unique in that regard.

Some background is helpful here:

* This datasource does not depend on L3 networking, even an ephemeral DHCP network, to fetch its network configuration. Instead this currently happens via a series of RPC calls for `guestinfo` key/value pairs.
* This datasource was created in an environment where DHCP was used, which was not initially a big deal, but...
* Eventually the need to support Cluster API (CAPI) meant supporting cloud-init instance data JINJA queries
* Except cloud-init instance data only includes the information passed _into_ the metadata
* This is why this datasource uses the `setup` function to inspect the network after it is online and repersists the metadata back out to disk as the instance data file
* Now the user and vendor data can execute JINJA queries in the network boot stage and beyond that discover hostname, local_ipv4, and other information that were not available until the VM's DHCP network was brought online in the local boot stage.

All this is to say the reason this datasource's `getfqdn` function does not look at the metadata is because:

* The result of `getfqdn` might be used to _update_ the metadata in order to persist the DHCP-provided hostname back to disk in its instance data
* The `get_hostname` function has comments that maybe it should return an error if the datasource doesn't provide a hostname when requested.

So to me it seems like this is behavior that should continue to exist in the datasource and not rely on an implementation of the base class that could change.

Reply:
There is a also a question of whether it might be better to use util.get_hostname() instead as well (it uses socket.gethostname() instead of talking the IMDS)